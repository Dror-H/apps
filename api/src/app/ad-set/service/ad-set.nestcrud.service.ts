import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetNestCrudService extends TypeOrmCrudService<AdSet> {
  constructor(@Inject(AdSetRepository) repo: AdSetRepository) {
    super(repo);
  }
}
