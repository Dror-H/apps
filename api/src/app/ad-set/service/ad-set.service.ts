import {
  AdSetCreationDTO,
  WorkspaceDTO,
  AdSetDTO,
  ResponseSuccess,
  SupportedProviders,
  TargetingDto,
  CacheTtlSeconds,
} from '@instigo-app/data-transfer-object';
import { ThirdPartyAdSetApiService } from '@instigo-app/third-party-connector';
import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';
import { groupBy } from 'lodash';
import { from, Observable } from 'rxjs';
import { map, mergeMap, switchMap } from 'rxjs/operators';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetService {
  @Inject(ThirdPartyAdSetApiService)
  private readonly thirdPartyAdSetApiService: ThirdPartyAdSetApiService;

  @Inject(AdSetRepository)
  private readonly repo: AdSetRepository;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(adSet: Partial<AdSetCreationDTO>, currentWorkspace: WorkspaceDTO): Promise<AdSetDTO> {
    const { provider, adAccount, campaign } = adSet;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAdSet = await this.thirdPartyAdSetApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      adSet: adSet,
    });
    builtAdSet.adAccount = adAccount;
    builtAdSet.campaign = campaign;
    return this.repo.save(builtAdSet);
  }

  bulkPatch(options: { adSets: Partial<AdSetDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { adSets, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(adSets, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdSetApiService.bulkPatch({ provider, accessToken, adSets: entries });
      }),
      switchMap(() => this.repo.save(adSets)),
      map(() => new ResponseSuccess('AdSets successfully updated')),
    );
  }

  async findAdSetTargeting(options: {
    workspace: WorkspaceDTO;
    provider: SupportedProviders;
    adSetId: string;
  }): Promise<TargetingDto> {
    const { adSetId, provider, workspace } = options;
    const cacheKey = `adSetTargeting-${provider}-${adSetId}`;
    let result = await this.cacheManager.get(cacheKey);
    if (result) {
      return result as any;
    }
    const { accessToken } = workspace.owner.getAccessToken({ provider });
    result = await this.thirdPartyAdSetApiService.findAdSetTargeting({
      accessToken,
      adSetId,
      provider,
    });
    await this.cacheManager.set(cacheKey, result, { ttl: CacheTtlSeconds.ONE_HOUR });
    return result as any;
  }
}
