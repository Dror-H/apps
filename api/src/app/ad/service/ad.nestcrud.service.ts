import { Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Ad } from '../data/ad.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

@Injectable()
export class AdNestCrudService extends TypeOrmCrudService<Ad> {
  constructor(
    @InjectRepository(Ad)
    public repo: Repository<Ad>,
  ) {
    super(repo);
  }
}
