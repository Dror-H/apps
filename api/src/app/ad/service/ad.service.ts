import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { AdCreationDTO, AdDTO, AdUpdateDTO, ResponseSuccess, WorkspaceDTO } from '@instigo-app/data-transfer-object';
import { Ad } from '@api/ad/data/ad.entity';
import { ThirdPartyAdApiService } from '@instigo-app/third-party-connector';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Cache } from 'cache-manager';
import { from, Observable } from 'rxjs';
import { groupBy } from 'lodash';
import { map, mergeMap, switchMap } from 'rxjs/operators';

@Injectable()
export class AdService {
  @Inject(ThirdPartyAdApiService)
  private readonly thirdPartyAdApiService: ThirdPartyAdApiService;

  @InjectRepository(Ad)
  private readonly adRepository: Repository<Ad>;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(ad: Partial<AdCreationDTO<any>>, currentWorkspace: WorkspaceDTO): Promise<AdDTO> {
    const { provider, adAccount } = ad;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAd = await this.thirdPartyAdApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ad,
    });
    builtAd.adAccount = adAccount;
    return this.adRepository.save(builtAd);
  }

  async bulkCreate(ads: Partial<AdCreationDTO<any>>[], currentWorkspace: WorkspaceDTO): Promise<Ad[]> {
    const { provider, adAccount, campaign } = ads[0];
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const createdAds = await this.thirdPartyAdApiService.createMany({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ads: ads,
    });
    const adSet = ads[0].providerSpecificFields?.adSet;
    const mappedAds = createdAds.map((item) => ({ ...item, adAccount, campaign, adSet }));
    return this.adRepository.save(mappedAds);
  }

  public async updateOne(ad: AdUpdateDTO, currentWorkspace: WorkspaceDTO): Promise<Ad> {
    await this.cacheManager.reset();
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider: ad.provider });
    await this.thirdPartyAdApiService.updateAd({ provider: ad.provider, accessToken, ad });
    const newAd = {
      id: ad.id,
      name: ad.name,
      objectStorySpec: { ...ad.data, adTemplateType: ad.adTemplateType },
    } as Ad;
    return await this.adRepository.save(newAd);
  }

  bulkPatch(options: { ads: Partial<AdDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { ads, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(ads, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdApiService.bulkPatch({ provider, accessToken, ads: entries });
      }),
      switchMap(() => this.adRepository.save(ads)),
      map(() => new ResponseSuccess('Ads successfully updated')),
    );
  }
}
