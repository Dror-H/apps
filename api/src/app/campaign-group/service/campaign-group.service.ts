import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { CampaignGroup } from '../data/campaign-group.entity';
import { CampaignGroupRepository } from '../data/campaign-group.repository';

@Injectable()
export class CampaignGroupNestCrudService extends TypeOrmCrudService<CampaignGroup> {
  constructor(@Inject(CampaignGroupRepository) public repo: CampaignGroupRepository) {
    super(repo);
  }
}
