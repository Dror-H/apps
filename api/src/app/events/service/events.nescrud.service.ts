import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { EventsRepository } from '@api/events/data/events.repository';
import { EventsEntity } from '@api/events/data/events.entity';

@Injectable()
export class EventsNestCrudService extends TypeOrmCrudService<EventsEntity> {
  constructor(@Inject(EventsRepository) eventsRepository: EventsRepository) {
    super(eventsRepository);
  }
}
