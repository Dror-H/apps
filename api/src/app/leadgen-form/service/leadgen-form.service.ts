import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { LeadgenForm } from '../data/leadgen-form.entity';
import { LeadgenFormRepository } from '../data/leadgen_form.repository';

@Injectable()
export class LeadgenFormNestCrudService extends TypeOrmCrudService<LeadgenForm> {
  constructor(@Inject(LeadgenFormRepository) public repo: LeadgenFormRepository) {
    super(repo);
  }
}
