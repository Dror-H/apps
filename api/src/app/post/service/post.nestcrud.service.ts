import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Post } from '../data/post.entity';
import { PostRepository } from '../data/post.repository';

@Injectable()
export class PostNestCrudService extends TypeOrmCrudService<Post> {
  constructor(@Inject(PostRepository) repo: PostRepository) {
    super(repo);
  }
}
