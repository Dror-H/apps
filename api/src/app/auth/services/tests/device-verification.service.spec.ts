import { UserDeviceMetadataRepository } from '@api/user/data/user-device-metadata.repository';
import { userMock } from '@instigo-app/api-shared';
import { Test } from '@nestjs/testing';
import { EventEmitter2 } from 'eventemitter2';
import { DeviceVerificationService } from '../device-verification.service';
import * as faker from 'faker';
describe('DeviceVerificationService Test suite', () => {
  let service: DeviceVerificationService;
  let mailerService;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        DeviceVerificationService,
        { provide: EventEmitter2, useValue: {} },
        { provide: UserDeviceMetadataRepository, useValue: {} },
        { provide: 'MailerService', useValue: {} },
      ],
    }).compile();
    mailerService = module.get('MailerService');
    service = module.get<DeviceVerificationService>(DeviceVerificationService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be created', () => {
    const user = userMock();
    user.email = 'Douglas.Farrell@gmail.com';
    user.fullName = 'Hipolito,Gerhold';
    const unknownIp = '233.201.170.95';
    const mail = service.emailBody({ user, unknownIp });
    expect(mail).toMatchSnapshot();
  });

  it('should not crash sent', async () => {
    const user = userMock();
    const unknownIp = faker.internet.ip();
    mailerService.sendMail = jest.fn().mockRejectedValue({});
    await expect(service.notifyUserByEmail({ user, unknownIp })).resolves.toBeUndefined();
  });
});
