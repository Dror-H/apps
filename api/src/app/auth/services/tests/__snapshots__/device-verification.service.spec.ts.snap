// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceVerificationService Test suite should be created 1`] = `
Object {
  "context": Object {
    "from": "233.201.170.95",
    "name": "Hipolito,Gerhold",
  },
  "subject": "Unknown device login - Instigo",
  "template": "unknown-device-login",
  "to": "Douglas.Farrell@gmail.com",
}
`;
